File: apps/tlon-mobile/android/app/src/main/java/io/tlon/landscape/models/Yarn.java
package io.tlon.landscape.models;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Optional;

public class Yarn {

    public Boolean isValidNotification;
    public Boolean isGroup;
    public Boolean isClub;
    public String wer;
    public Optional<String> channelId;
    public Optional<String> groupId;
    public String senderId;
    public String contentText;

    public Yarn(JSONObject yarnObject) throws JSONException {
        JSONObject rope = yarnObject.getJSONObject("rope");
        String thread = rope.getString("thread");
        JSONArray content = yarnObject.getJSONArray("con");
        ArrayList<String> parts = new ArrayList<>();
        boolean skipNextPart = false;
        for (int i = 0; i < content.length(); i++) {
            if (skipNextPart) {
                skipNextPart = false;
                continue;
            }

            Object item = content.get(i);
            if (item instanceof JSONObject) {
                JSONObject itemObj = (JSONObject) item;
                if (itemObj.has("ship")) {
                    String ship = itemObj.getString("ship");
                    if (senderId == null) {
                        senderId = ship;
                    }
                } else if (itemObj.has("emph")) {
                    parts.add(itemObj.getString("emph"));
                }
            } else if (item instanceof String) {
                String text = (String) item;
                boolean isReply = text.equals(" replied to ") // DM replies
                        || text.equals(" replied to your message “"); // Backwards compatibility
                if (isReply
                    || text.equals(": ")
                    || text.equals("”: ")
                    || text.equals(" sent a message: ") // User has all notifications enabled
                    || text.equals(" replied to you: ") // Group replies, don't skip next
                    || text.equals(" mentioned you: ") // Mentions
                    || text.equals(" mentioned you:") // Backwards compatibility
                    || text.equals(" mentioned you :") // Backwards compatibility
                ) {
                    skipNextPart = isReply;
                    continue;
                }

                parts.add(text);
            }
        }

        isGroup = !rope.isNull("group");
        isValidNotification = rope.getString("desk").equals("groups") &&
                (!thread.endsWith("/channel/edit") &&
                !thread.endsWith("/channel/add") &&
                !thread.endsWith("/channel/del") &&
                !thread.endsWith("/joins") &&
                !thread.endsWith("/leaves") ||
                !isGroup);
        wer = yarnObject.getString("wer");
        groupId = !isGroup ? Optional.empty() :
                Optional.of(rope.getString("group"));
        channelId = isGroup ?
                rope.isNull("channel") ? Optional.empty() : Optional.of(rope.getString("channel"))
                : Optional.of(thread.replace("/dm/", "").replace("/club/", ""));
        isClub = channelId.isPresent() && channelId.get().startsWith("0v");
        contentText = String.join("", parts);
        if (senderId == null && channelId.isPresent() && channelId.get().startsWith("~")) {
            senderId = channelId.get();
        }
    }

}

File: apps/tlon-mobile/ios/Shared/Models/Yarn.swift
//
//  Yarn.swift
//  Pocket
//
//  Created by Alec Ananian on 8/5/22.
//

import Foundation
import NotificationCenter

struct Yarn: Decodable {
    let id: String
    let rope: Rope
    let time: UInt
    let con: [YarnContent]
    let wer: String

    // Flag group invitations and group-meta events as invalid
    var isValidNotification: Bool {
        (rope.desk == "groups") &&
            (!rope.thread.hasSuffix("/channel/edit") &&
            !rope.thread.hasSuffix("/channel/add") &&
            !rope.thread.hasSuffix("/channel/del") &&
            !rope.thread.hasSuffix("/joins") &&
            !rope.thread.hasSuffix("/leaves") || !isGroup)
    }

    var isGroup: Bool {
        rope.group != nil
    }

    var isClub: Bool {
      (channelID ?? "").starts(with: "0v")
    }

    var isInvitation: Bool {
        con.contains(where: { $0.text == " has invited you to a direct message" })
    }

    var category: NotificationCategory {
        isInvitation ? .invitation : .message
    }

    var channelID: String? {
        if isGroup && rope.channel == nil {
          return nil
        }
      
        if isGroup, let channel = rope.channel {
            return channel
        }

        return rope.thread.replacingOccurrences(of: "/dm/", with: "").replacingOccurrences(of: "/club/", with: "")
    }
  
    var groupID: String? {
        return rope.group
    }

    var senderShipName: String? {
        for item in con {
            if case let .ship(result) = item {
                return result.ship
            }
        }

        return nil
    }

    var body: String {
        var parts = [String]()
        var skipNextPart = false
        con.filter { !$0.isShipName }.forEach { content in
            if skipNextPart {
                skipNextPart = false
                return
            }

            let text = content.text
            let isReply = text == " replied to " // DM replies
                || text == " replied to your message “" // Backwards compatibility
            if isReply
                || text == ": "
                || text == "”: "
                || text == " sent a message: " // User has all notifications enabled
                || text == " replied to you: " // Group replies, don't skip next
                || text == " mentioned you: " // Mentions
                || text == " mentioned you:" // Backwards compatibility
                || text == " mentioned you :" // Backwards compatibility
            {
                skipNextPart = isReply
                return
            }

            parts.append(text)
        }

        return parts.joined(separator: "")
    }

    var userInfo: [AnyHashable: Any] {
        [
            "wer": wer,
            "channelId": channelID,
        ]
    }

    func getTitle() async -> String {
        // Full message will go in body
        if isInvitation {
            return ""
        }

        // Sender name will be added to intent
        if !isGroup, !isClub {
            return ""
        }

        var displayName: String? = channelID ?? groupID

        if let channelID {
          if isGroup {
            do {
              let groupChannel = try await GroupChannelStore.sharedInstance.getOrFetchItem(channelID)
              displayName = groupChannel?.meta.title
            } catch {
              error.logWithDomain(TlonError.NotificationsFetchGroupChannel)
            }
          } else {
            do {
              let club = try await ClubStore.sharedInstance.getOrFetchItem(channelID)
              displayName = club?.displayName
            } catch {
              error.logWithDomain(TlonError.NotificationsFetchClub)
            }
          }
        } else if let groupID {
            do {
              let group = try await GroupStore.sharedInstance.getOrFetchItem(groupID)
              displayName = group?.preview?.meta.title
            } catch {
              error.logWithDomain(TlonError.NotificationsFetchGangs)
            }
        }

        return displayName ?? ""
    }
}

struct YarnContentShip: Decodable {
    let ship: String
}

struct YarnContentEmphasis: Decodable {
    let emph: String
}

indirect enum YarnContent: Decodable {
    case text(String)
    case ship(YarnContentShip)
    case emph(YarnContentEmphasis)

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()

        if let value = try? container.decode(String.self) {
            self = .text(value)
        } else if let value = try? container.decode(YarnContentShip.self) {
            self = .ship(value)
        } else if let value = try? container.decode(YarnContentEmphasis.self) {
            self = .emph(value)
        } else {
            throw DecodingError.typeMismatch(
                YarnContent.self,
                DecodingError.Context(codingPath: decoder.codingPath, debugDescription: "Wrong type for YarnContent")
            )
        }
    }

    var text: String {
        switch self {
        case let .text(value):
            return value

        case let .ship(value):
            return ContactStore.sharedInstance.getItem(value.ship)?.displayName ?? value.ship

        case let .emph(value):
            return value.emph
        }
    }

    var isShipName: Bool {
        if case .ship = self {
            return true
        }

        return false
    }
}

struct Rope: Decodable {
    let group: String?
    let channel: String?
    let desk: String
    let thread: String
}
